packages/sdk/src/evm/core/classes/transactions.ts

```diff
@@ -0,0 +1,458 @@
+import { TransactionError, parseRevertReason } from "../../common/error";
+import { getPolygonGasPriorityFee } from "../../common/gas-price";
+import {
+  fetchContractMetadataFromAddress,
+  fetchSourceFilesFromMetadata,
+} from "../../common/metadata-resolver";
+import { isBrowser } from "../../common/utils";
+import { ChainId } from "../../constants/chains";
+import { ContractSource } from "../../schema/contracts/custom";
+import { TransactionResult } from "../types";
+import { ContractWrapper } from "./contract-wrapper";
+import { ConnectionInfo } from "@ethersproject/web";
+import { ThirdwebStorage } from "@thirdweb-dev/storage";
+import { BaseContract, BigNumber, CallOverrides, ethers } from "ethers";
+
+type TransactionOptions = {
+  method: string;
+  args: any[];
+  overrides?: CallOverrides;
+  storage?: ThirdwebStorage;
+};
+
+type TransactionOptionsWithContractWrapper<
+  TContract extends BaseContract = BaseContract,
+> = TransactionOptions & {
+  contractWrapper: ContractWrapper<TContract>;
+};
+
+type TransactionOptionsWithContract = Omit<TransactionOptions, "contract"> & {
+  contract: ethers.Contract;
+  provider: ethers.providers.Provider;
+  signer: ethers.Signer;
+};
+
+type TransactionOptionsWithContractInfo = Omit<
+  TransactionOptionsWithContract,
+  "contract"
+> & {
+  provider: ethers.providers.Provider;
+  contractAddress: string;
+  contractAbi?: ethers.ContractInterface;
+};
+
+export function transaction(fn: (...args: any[]) => Promise<Transaction>) {
+  async function executeFn(...args: any[]): Promise<TransactionResult> {
+    const tx = await fn(...args);
+    return { receipt: await tx.execute() };
+  }
+
+  executeFn.transaction = fn;
+  return executeFn;
+}
+
+export class Transaction {
+  private contract: ethers.Contract;
```

Comment:
Need a good pattern to expose these (especially method/args/overrides). Maybe `get` or make them public. Alternatively a function for each like `.args()`